### ISSUE TITLE
Load a small subset of data from a big dataset takes forever

### ISSUE THREAD

**Issue by malmans2** [pydata/xarray#1985]
#### Code Sample
```python
def cut_dataset(ds2cut, 
                varList    = ['Temp' 'S' 'Eta' 'U' 'V' 'W'], 
                lonRange   = [-180, 180],
                latRange   = [-90, 90],
                depthRange = [0, float("inf")],
                timeRange  = ['2007-09-01T00', '2008-08-31T18'],
                timeFreq   = '1D',
                sampMethod = 'mean',
                interpC    = True,
                saveNetCDF = False):
    """
    Cut the dataset
    """
    
    # Copy dataset
    ds = ds2cut.copy(deep=True)

    # Choose variables
    varList_tmp = list(varList)
    for varName in ds.variables:
        if all(x != 'time' for x in ds[varName].dims) or (varName=='time'):
                varList_tmp.append(varName)
    toDrop = list(set(ds.variables)-set(varList_tmp))
    ds = ds.drop(toDrop)
    
    # Cut dataset
    ds = ds.sel(time = slice(min(timeRange),  max(timeRange)),
                Xp1  = slice(min(lonRange),   max(lonRange)),
                Yp1  = slice(min(latRange),   max(latRange)),
                Zp1  = slice(min(depthRange), max(depthRange)))
    ds = ds.sel(X    = slice(min(ds['Xp1'].values), max(ds['Xp1'].values)),
                Y    = slice(min(ds['Yp1'].values), max(ds['Yp1'].values)),
                Z    = slice(min(ds['Zp1'].values), max(ds['Zp1'].values)),
                Zu   = slice(min(ds['Zp1'].values), max(ds['Zp1'].values)),
                Zl   = slice(min(ds['Zp1'].values), max(ds['Zp1'].values)))

    # Resample
    if sampMethod=='snapshot':
        ds = ds.resample(time=timeFreq).first(skipna=False)
    elif sampMethod=='mean':
        ds = ds.resample(time=timeFreq).mean()
        
    # Create grid
    grid = xgcm.Grid(ds,periodic=False)
    
    # Interpolate
    if interpC:
        for varName in varList:
            for dim in ds[varName].dims:
                if len(dim)>1 and dim!='time':
                    ds[varName] = grid.interp(ds[varName], axis=dim[0])

    # Remove useless variables
    allDims = []
    for varName in varList:
        for dim in ds[varName].dims:
            allDims.append(dim)
    toDrop = []
    for varName in ds.variables:
        if len(list(set(ds[varName].dims)-set(allDims)))>0:
            toDrop.append(varName)
    ds = ds.drop(toDrop)
    
    # Save to NetCDF
    if saveNetCDF: ds.to_netcdf(saveNetCDF)
    
    return ds, grid

# 3D test
ds_cut, grid_cut = cut_dataset(ds, 
                               varList    = ['Eta'], 
                               latRange   = [65, 65.5],
                               depthRange = [0, 2],
                               timeRange  = ['2007-11-15T00', '2007-11-16T00'],
                               timeFreq   = '1D',
                               sampMethod = 'mean',
                               interpC    = False,
                               saveNetCDF = '3Dvariable.nc')

# 4D test
ds_cut, grid_cut = cut_dataset(ds, 
                               varList    = ['Temp'], 
                               lonRange   = [-30, -29.5],
                               latRange   = [65, 65.5],
                               depthRange = [0, 2],
                               timeRange  = ['2007-11-15T00', '2007-11-16T00'],
                               timeFreq   = '1D',
                               sampMethod = 'mean',
                               interpC    = False,
                               saveNetCDF = '4Dvariable.nc')
```
#### Problem description

I'm working with a big dataset. However, most of the time I only need a small subset of data. My idea was to open and concatenate everything with open_mfdataset, and then extract subsets of data using the indexing routines. This approach works very good when I extract 3D variables (just lon, lat, and time), but it fails when I try to extract 4D variables (lon, lat, time, and depth). It doesn't actually fail, but to_netcdf takes forever. When I open a smaller dataset since the very beginning (let's say just November), then I'm also able to extract 4D variables. 
When I load the sub-dataset after using the indexing routines, does xarray need to read the whole original 4D variable? If yes, then I should probably change my approach and I should open subset of data since the very beginning. If no, am I doing something wrong?

#### Output of ``xr.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-327.18.2.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

xarray: 0.10.1
pandas: 0.20.1
numpy: 1.11.0
scipy: 0.17.1
netCDF4: 1.3.1
h5netcdf: 0.5.0
h5py: 2.7.1
Nio: None
zarr: None
bottleneck: 1.2.1
cyordereddict: None
dask: 0.15.2
distributed: 1.18.1
matplotlib: 1.5.1
cartopy: 0.16.0
seaborn: None
setuptools: 27.2.0
pip: 8.1.2
conda: 4.4.11
pytest: 2.9.1
IPython: 4.2.0
sphinx: 1.4.1

</details>

**Comment by fujiisoup** [comment]
> When I load the sub-dataset after using the indexing routines, does xarray need to read the whole original 4D variable?

I don't think so. We support lazy indexing for any dimensional arrays (but not coordinate variables).
What does your data (especially '4Dvariable.nc') look like?
Is `Xp1` coordinate or sufficiently small?
`ds['Xp1'].values` loads `Xp1` into the memory.

**Comment by fujiisoup** [comment]
I notice this line
```python
    # Copy dataset
    ds = ds2cut.copy(deep=True)
```
loads the data into memory.
I think you don't need copy the dataset.

**Comment by malmans2** [comment]
I have the same issue if I don't copy the dataset.

Here are the coordinates of my dataset:
```python
<xarray.Dataset>
Dimensions:     (X: 960, Xp1: 961, Y: 880, Yp1: 881, Z: 216, Zl: 216, Zp1: 217, Zu: 216, time: 1465)
Coordinates:
  * Z           (Z) float64 1.0 3.5 7.0 11.5 17.0 23.5 31.0 39.5 49.0 59.5 ...
  * Zp1         (Zp1) float64 0.0 2.0 5.0 9.0 14.0 20.0 27.0 35.0 44.0 54.0 ...
  * Zu          (Zu) float64 2.0 5.0 9.0 14.0 20.0 27.0 35.0 44.0 54.0 65.0 ...
  * Zl          (Zl) float64 0.0 2.0 5.0 9.0 14.0 20.0 27.0 35.0 44.0 54.0 ...
  * X           (X) float64 -46.92 -46.83 -46.74 -46.65 -46.57 -46.48 -46.4 ...
  * Y           (Y) float64 56.81 56.85 56.89 56.93 56.96 57.0 57.04 57.08 ...
  * Xp1         (Xp1) float64 -46.96 -46.87 -46.78 -46.7 -46.61 -46.53 ...
  * Yp1         (Yp1) float64 56.79 56.83 56.87 56.91 56.95 56.98 57.02 ...
  * time        (time) datetime64[ns] 2007-09-01 2007-09-01T06:00:00 ...
```
I don't think the horizontal coordinates are the problem because it works fine when I use the same function on 3D variables.
I'm also attaching the function that I use to open the dataset, just in case is helpful:
```python
def load_dataset():
    """
    Load the whole dataset
    """
    
    # Import grid and fields separately, then merge
    gridpath = '/home/idies/workspace/OceanCirculation/exp_ASR/grid_glued.nc'
    fldspath = '/home/idies/workspace/OceanCirculation/exp_ASR/result_*/output_glued/*.*_glued.nc'
    gridset = xr.open_dataset(gridpath,
                              drop_variables = ['XU','YU','XV','YV','RC','RF','RU','RL'])
    fldsset = xr.open_mfdataset(fldspath,
                                concat_dim     = 'T',
                                drop_variables = ['diag_levels','iter'])
    ds = xr.merge([gridset, fldsset])
    
    # Adjust dimensions creating conflicts
    ds = ds.rename({'Z': 'Ztmp'})
    ds = ds.rename({'T': 'time', 'Ztmp': 'Z', 'Zmd000216': 'Z'})
    ds = ds.squeeze('Zd000001')
    for dim in ['Z','Zp1', 'Zu','Zl']:
        ds[dim].values   = np.fabs(ds[dim].values)
        ds[dim].attrs.update({'positive': 'down'})

    # Create horizontal vectors (remove zeros due to exch2)
    ds['X'].values   = ds.XC.where((ds.XC!=0) & (ds.YC!=0)).mean(dim='Y',   skipna=True)
    ds['Xp1'].values = ds.XG.where((ds.XG!=0) & (ds.YG!=0)).mean(dim='Yp1', skipna=True)
    ds['Y'].values   = ds.YC.where((ds.XC!=0) & (ds.YC!=0)).mean(dim='X',   skipna=True)
    ds['Yp1'].values = ds.YG.where((ds.XG!=0) & (ds.YG!=0)).mean(dim='Xp1', skipna=True)
    ds = ds.drop(['XC','YC','XG','YG'])    
        
    # Create xgcm grid
    ds['Z'].attrs.update({'axis': 'Z'})
    ds['X'].attrs.update({'axis': 'X'})
    ds['Y'].attrs.update({'axis': 'Y'})
    for dim in ['Zp1','Zu','Zl','Xp1','Yp1']:
        if min(ds[dim].values)<min(ds[dim[0]].values):
            ds[dim].attrs.update({'axis': dim[0], 'c_grid_axis_shift': -0.5})
        elif min(ds[dim].values)>min(ds[dim[0]].values):
            ds[dim].attrs.update({'axis': dim[0], 'c_grid_axis_shift': +0.5})
    grid = xgcm.Grid(ds,periodic=False)
        
    return ds, grid
```

**Comment by fujiisoup** [comment]
Umm. I could not find what is wrong with your code.
Can you find which line loads the data into memory?
If your data is still a dask array, it does not print the entries of the array but instead, it shows something like this,
```python
<xarray.DataArray (x: 3)>
dask.array<shape=(3,), dtype=int64, chunksize=(3,)>
Dimensions without coordinates: x
```

**Comment by malmans2** [comment]
The problem occurs when I run the very last line, which is to_netcdf().
Right before, the dataset looks like this:
```python
<xarray.Dataset>
Dimensions:  (X: 10, Y: 25, Z: 1, time: 2)
Coordinates:
  * time     (time) datetime64[ns] 2007-11-15 2007-11-16
  * Z        (Z) float64 1.0
  * X        (X) float64 -29.94 -29.89 -29.85 -29.81 -29.76 -29.72 -29.67 ...
  * Y        (Y) float64 65.01 65.03 65.05 65.07 65.09 65.11 65.13 65.15 ...
Data variables:
    drF      (time, Z) float64 2.0 2.0
    dxF      (time, Y, X) float64 2.066e+03 2.066e+03 2.066e+03 2.066e+03 ...
    dyF      (time, Y, X) float64 2.123e+03 2.123e+03 2.123e+03 2.123e+03 ...
    rA       (time, Y, X) float64 4.386e+06 4.386e+06 4.386e+06 4.386e+06 ...
    fCori    (time, Y, X) float64 0.0001322 0.0001322 0.0001322 0.0001322 ...
    R_low    (time, Y, X) float64 -2.001e+03 -1.989e+03 -1.973e+03 ...
    Ro_surf  (time, Y, X) float64 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ...
    Depth    (time, Y, X) float64 2.001e+03 1.989e+03 1.973e+03 1.963e+03 ...
    HFacC    (time, Z, Y, X) float64 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 ...
    Temp     (time, Z, Y, X) float64 dask.array<shape=(2, 1, 25, 10), chunksize=(1, 1, 25, 10)>
```
This is a dask array, right?

**Comment by malmans2** [comment]
I forgot to mention that I'm getting this warning:
/home/idies/anaconda3/lib/python3.5/site-packages/dask/core.py:306: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  elif type_arg is type(key) and arg == key:

However, I don't think it is relevant since I get the same warning when I'm able to run .to_netcdf() on the 3D variable.

**Comment by fujiisoup** [comment]
I see no problem with your code...
Can you try updating xarray to 0.10.2 (released today)? We updated some logic of lazy indexing.

**Comment by malmans2** [comment]
Alright, I found the problem.
I'm loading several variables from different files.
All the variables have 1464 snapshots. However, one of the 3D variables has just one snapshot at a different time (I found a bag in my bash script to re-organize the raw data). When I load my dataset using .open_mfdataset, the time dimension has an extra snapshot (length is 1465). However, xarray doesn't like it and when I run functions such as to_netcdf it takes forever (no error).
Thanks @fujiisoup for the help!
